cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/ServiceComponentProgramRunner.java

```diff
@@ -119,19 +123,20 @@ public ProgramController run(Program program, ProgramOptions options) {
       component = new ServiceWorkerDriver(program, workerSpec, context);
     }
 
-    ProgramControllerServiceAdapter controller = new ProgramControllerServiceAdapter(component, componentName, runId);
+    ProgramControllerServiceAdapter controller =
+      new ServiceComponentProgramControllerAdapter(component, componentName, runId);
     component.start();
     return controller;
   }
 
-  private BasicHttpServiceContextFactory createHttpServiceContextFactory(final Program program,
-                                                                         final RunId runId,
+  private BasicHttpServiceContextFactory createHttpServiceContextFactory(final Program program, final RunId runId,
```

Comment:
It's actually better in your last commit that this method get moved into ServiceHttpServer so that you don't split the `AtomicInteger` into two places. If this method is inside ServiceHttpServer, it can access the instanceCount field directly